### PR TITLE
fix(Radio+Checkbox): use input component

### LINKED ISSUE #2459: Topp align checkbox and radio
Issue reported by team in Mattilsynet:Image

### PR DESCRIPTION
**Pull request by eirikbacker**
- Part of #2311
- Fieldset at compound components moved to task #2666 
- Fixes #2459

### CODE REVIEW

File: packages/react/src/components/form/Checkbox/Checkbox.tsx [hunk 1]
```diff
@@ -5,90 +5,52 @@ import { forwardRef } from 'react';
...
+  function Checkbox(
+    { children, label, description, validation, ...rest },
+    ref,
+  ) {
+    // const inputRef = useMergeRefs<HTMLInputElement>([
+    //   ref,
+    //   (el) => {
+    //     if (el) el.indeterminate = rest.indeterminate ?? false;
+    //   },
+    // ]);
```
Comment (eirikbacker) [inline packages/react/src/components/form/Checkbox/Checkbox.tsx:45]: Should this be in input instead? 🤔 Or simply remove, and instead show in a story, as this prevents server component behavior?

File: packages/css/fieldset.css [hunk 2]
```diff
@@ -1,44 +1,23 @@
...
 
-  &[data-readonly] > legend::before {
-    content: '';
-    background: currentcolor;
-    height: var(--dsc-fieldset-icon-size);
-    mask: center / contain no-repeat
-      url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' aria-hidden='true' viewBox='0 0 24 24'%3E%3Cpath fill-rule='evenodd' d='M12 2.25A4.75 4.75 0 0 0 7.25 7v2.25H7A1.75 1.75 0 0 0 5.25 11v9c0 .41.34.75.75.75h12a.75.75 0 0 0 .75-.75v-9A1.75 1.75 0 0 0 17 9.25h-.25V7A4.75 4.75 0 0 0 12 2.25m3.25 7V7a3.25 3.25 0 0 0-6.5 0v2.25zM12 13a1.5 1.5 0 0 0-.75 2.8V17a.75.75 0 0 0 1.5 0v-1.2A1.5 1.5 0 0 0 12 13'/%3E%3C/svg%3E");
-    width: var(--dsc-fieldset-icon-size);
-  }
+  & > p {
```
Comment (eirikbacker) [inline packages/css/fieldset.css:14]: Should fieldset style `p`, or `[data-field="description"]`? 🤔

File: packages/css/fieldset.css [hunk 3]
```diff
@@ -39,6 +24,6 @@
 
   &:disabled > legend,
   &:disabled > legend + p {
```
Comment (eirikbacker) [inline packages/css/fieldset.css:26]: Design: Are we sure about this?

File: packages/css/label.css [hunk 4]
```diff
@@ -1,12 +1,24 @@
 .ds-label {
-  /* We set these three styles when label is used in `legend` */
-  display: inline-block;
-  margin: 0;
-  padding: 0;
-
-  @composes ds-label--md from './base/base.css';
+  --dsc-label--readonly: initial; /* Using technique https://css-tricks.com/the-css-custom-property-toggle-trick/ */
```
Comment (eirikbacker) [inline packages/css/label.css:2]: Should label also be exposed as `<Legend>`? 🤔

File: packages/react/src/components/form/Fieldset/Fieldset.tsx [hunk 5]
```diff
@@ -57,7 +57,6 @@ export const Fieldset = forwardRef<HTMLFieldSetElement, FieldsetProps>(
           aria-invalid={inputProps['aria-invalid']}
           className={cl('ds-fieldset', className)}
           data-hidelegend={hideLegend || undefined}
```
Comment (eirikbacker) [inline packages/react/src/components/form/Fieldset/Fieldset.tsx:59]: Remove hidelegend?

File: packages/react/src/components/form/Checkbox/Checkbox.tsx [hunk 6]
```diff
@@ -5,90 +5,52 @@ import { forwardRef } from 'react';
...
+import { Field } from '../Field';
+import { Input } from '../Input';
 import type { FormFieldProps } from '../useFormField';
 
 import { useCheckbox } from './useCheckbox';
 
 export type CheckboxProps = {
   /** Checkbox label */
-  children?: ReactNode;
+  label?: ReactNode;
```
Comment (eirikbacker) [inline packages/react/src/components/form/Checkbox/Checkbox.tsx:17]: Should have **label** _or_ **aria-label** required

File: packages/react/src/components/form/Checkbox/Checkbox.tsx [hunk 7]
```diff
@@ -5,90 +5,52 @@ import { forwardRef } from 'react';
...
-              )}
-            </>
-          )}
-        </div>
-      </Paragraph>
+      <Field>
+        <Input type='checkbox' ref={ref} {...rest} />
+        {label && <Label weight='regular'>{label}</Label>}
+        {description && <div data-field='description'>{description}</div>}
+        {validation && <ValidationMessage>{validation}</ValidationMessage>}
```
Comment (eirikbacker) [inline packages/react/src/components/form/Checkbox/Checkbox.tsx:52]:
```suggestion
        {!!label && <Label weight='regular'>{label}</Label>}
        {!!description && <div data-field='description'>{description}</div>}
        {!!validation && <ValidationMessage>{validation}</ValidationMessage>}
```

File: packages/react/src/components/form/Checkbox/CheckboxGroup.stories.tsx [hunk 8]
```diff
@@ -35,98 +44,110 @@ Preview.args = {
...
+    setValues(
+      values.includes(value)
+        ? values.filter((val) => val !== value)
+        : [...values, value],
     );
   return (
     <>
-      <Checkbox.Group
+      <Fieldset
         legend='Skal du reise til noen av disse landene?'
```
Comment (eirikbacker) [inline packages/react/src/components/form/Checkbox/CheckboxGroup.stories.tsx:69]: Fieldset.Legend

File: packages/react/src/components/form/Radio/RadioGroup.tsx [hunk 9]
```diff
@@ -71,11 +71,9 @@ export const RadioGroup = forwardRef<HTMLFieldSetElement, RadioGroupProps>(
           }}
         >
           <div
-            className={cl(
-              'ds-radio__group',
-              `ds-radio__group--${size}`,
-              inline && 'ds-radio__group--horizontal',
-            )}
+            className={cl('ds-radio__group', className)}
+            data-inline={inline || undefined}
```
Comment (eirikbacker) [inline packages/react/src/components/form/Radio/RadioGroup.tsx:75]: Remove inline, and instead show example with `display: flex`

File: packages/react/src/components/form/Checkbox/CheckboxGroup.stories.tsx [hunk 10]
```diff
@@ -35,98 +44,110 @@ Preview.args = {
...
 export const Controlled: StoryFn<typeof Checkbox> = () => {
-  const [value, setValue] = useState<string[]>([]);
-
-  const myToggle = (val: string) =>
-    setValue(
-      value.includes(val) ? value.filter((x) => x !== val) : [...value, val],
+  const [values, setValues] = useState<string[]>([]);
+
+  const onChange = (event: ChangeEvent<HTMLInputElement>) =>
+    setValues(Checkbox.getValues(event));
```
Comment (eirikbacker) [inline packages/react/src/components/form/Checkbox/CheckboxGroup.stories.tsx:59]: `useCheckbox`

File: packages/react/src/components/form/Checkbox/Checkbox.tsx (hunk 1 above)
Comment (eirikbacker) [reply packages/react/src/components/form/Checkbox/Checkbox.tsx:45]: Or move to hook?

File: packages/css/field.css [hunk 11]
```diff
@@ -1,7 +1,67 @@
...
+    --dsc-label--readonly: ; /* Activate lock icon for label when readonly */
+  }
+
+  /**
+   * Toggle inputs
+   */
+  &:has(input:is([type='radio'], [type='checkbox'])) {
+    border-radius: var(--ds-border-radius-md);
+    display: grid;
+    grid-template-columns: auto 1fr;
```
Comment (mimarz) [inline packages/css/field.css:39]: `grid-template-columns` and `focus` on checkboxes with `aria-label/describedby `🙄.
 
image

Comment (eirikbacker) [reply packages/css/field.css:39]: Oh! very nice catch! Will look into it

Comment (eirikbacker) [reply packages/css/field.css:39]: Fixed ✅

**Review by Barsnes — APPROVED** [review]

**Pull request merged**